cmd/fs-v1-background-append.go

```diff
@@ -0,0 +1,212 @@
+	info, ok := b.infoMap[uploadID]
+	if !ok {
+		return
+	}
+	delete(b.infoMap, uploadID)
+	close(info.endCh)
+}
+
+// This is run as a go-routine that appends the parts in the background.
+func (b *backgroundAppend) appendParts(disk StorageAPI, bucket, object, uploadID string, info bgAppendPartsInfo) {
+	// Holds the list of parts that is already appended to the "append" file.
+	appendMeta := fsMetaV1{}
+	for {
+		select {
+		case input := <-info.inputCh:
+			// We receive on this channel when new part gets uploaded or when complete-multipart sends
+			// a value on this channel to confirm if all the required parts are appended.
+			meta := input.meta
+			for {
+				// Append should be done such a way that if part-3 and part-2 is uploaded before part-1, we
+				// wait till part-1 is uploaded after which we append part-2 and part-3 as well in this for-loop.
+				part, appendNeeded := partToAppend(meta, appendMeta)
+				if !appendNeeded {
+					if reflect.DeepEqual(meta.Parts, appendMeta.Parts) {
+						// Sending nil is useful so that the complete-multipart-upload knows that
+						// all the required parts have been appended.
+						input.errCh <- nil
+					} else {
+						// Sending error is useful so that complete-multipart-upload can fall-back to
+						// its own append process.
+						input.errCh <- errPartsMissing
+					}
+					break
+				}
+				if err := appendPart(disk, bucket, object, uploadID, part); err != nil {
+					input.errCh <- err
+					break
+				}
+				appendMeta.AddObjectPart(part.Number, part.Name, part.ETag, part.Size)
+			}
+		case <-info.endCh:
+			// Either complete-multipart-upload or abort-multipart-upload closed endCh to end the appendParts go-routine.
+			appendFilePath := getFSAppendDataPath(uploadID)
+			disk.DeleteFile(bucket, appendFilePath) // Delete the tmp append file in case abort-multipart was called.
+			return
+		case <-time.After(appendPartsTimeout):
+			// Timeout the goroutine to garbage collect its resources. This would happen if the client initates
+			// a multipart upload and does not complete/abort it.
+			b.Lock()
+			delete(b.infoMap, uploadID)
+			b.Unlock()
+			// Delete the temporary append file as well.
+			appendFilePath := getFSAppendDataPath(uploadID)
+			disk.DeleteFile(bucket, appendFilePath)
+
+			close(info.timeoutCh)
+		}
+	}
+}
+
+// Appends the "part" to the append-file inside "tmp/" that finally gets moved to the actual location
+// upon complete-multipart-upload.
+func appendPart(disk StorageAPI, bucket, object, uploadID string, part objectPartInfo) error {
+	partPath := pathJoin(mpartMetaPrefix, bucket, object, uploadID, part.Name)
+	appendFilePath := getFSAppendDataPath(uploadID)
+
+	offset := int64(0)
+	totalLeft := part.Size
+	buf := make([]byte, readSizeV1)
+	for totalLeft > 0 {
+		curLeft := int64(readSizeV1)
+		if totalLeft < readSizeV1 {
+			curLeft = totalLeft
+		}
+		var n int64
+		n, err := disk.ReadFile(minioMetaBucket, partPath, offset, buf[:curLeft])
+		if n > 0 {
```

Comment:
we had discussed that the contract semantics of `disk.ReadFile()` will be same as `io.ReadFull()` so the consumer of this API can safely assume it and this code should always work. Code elsewhere also has been written based on this assumption. Should be put in the comment of `posix.ReadFile`

Also realized that ErrUnexpectedEOF should never happen as we always know what size to read.

Since it is more readable to eliminate error conditions first (as you mentioned) re-writing this to:

```
func appendPart(disk StorageAPI, bucket, object, uploadID string, part objectPartInfo) error {                                                                                  
        partPath := pathJoin(mpartMetaPrefix, bucket, object, uploadID, part.Name)                                                                                              
        appendFilePath := getFSAppendDataPath(uploadID)                                                                                                                         
                                                                                                                                                                                
        offset := int64(0)                                                                                                                                                      
        totalLeft := part.Size                                                                                                                                                  
        buf := make([]byte, readSizeV1)                                                                                                                                         
        for totalLeft > 0 {                                                                                                                                                     
                curLeft := int64(readSizeV1)                                                                                                                                    
                if totalLeft < readSizeV1 {                                                                                                                                     
                        curLeft = totalLeft                                                                                                                                     
                }                                                                                                                                                               
                var n int64                                                                                                                                                     
                n, err := disk.ReadFile(minioMetaBucket, partPath, offset, buf[:curLeft])                                                                                       
                if err == io.EOF {                                                                                                                                              
                        break                                                                                                                                                   
                }                                                                                                                                                               
                if err != nil {                                                                                                                                                 
                        disk.DeleteFile(bucket, appendFilePath)                                                                                                                 
                        return err                                                                                                                                              
                }                                                                                                                                                               
                if err = disk.AppendFile(minioMetaBucket, appendFilePath, buf[:n]); err != nil {                                                                                
                        disk.DeleteFile(bucket, appendFilePath)                                                                                                                 
                        return err                                                                                                                                              
                }                                                                                                                                                               
                offset += n                                                                                                                                                     
                totalLeft -= n                                                                                                                                                  
        }                                                                                                                                                                       
        return nil                                                                                                                                                              
}
```